add policy reference docs (#1199)
linkerd/website: linkerd.io/content/2.11/reference/policy-resources.md

+++

> title = "Server and ServerAuthroization"

Reviewer: speling

Reviewer 2: ```suggestion
title = "Server and ServerAuthorization"
```

description = "Details on the specification and what is possible with policy resources."
+++

[Server](#Server) and [ServerAuthorization](#ServerAuthorization) are the two types
of policy resources in Linkerd, used to control inbound access to your meshed
applications.

## Server

> [Server](#Server) provides a way to group your application pods.

Reviewer: How about this: A Server defines a port, selected across a set of pods, that is subject to traffic policy. Note that a Server can select pods across multiple workloads in a given namespace; for example, the admin port across every pod in the namespace. Note also that creating a Server object denies all communication to that port, and traffic will only be allowed if one or more corresponding `ServerAuthorization` objects exist.

Author: Better for sure!

### Spec

> A server spec must contain the following top level fields:

Reviewer: Capitalize Server

Reviewer 2: ```suggestion
A `Server` spec must contain the following top level fields:
```

{{< table >}}
| field| value |
|------|-------|

> | `podSelector`| a [podSelector](#podSelector) selects pods in the same namespace. |

Reviewer: Capitalize "A" in this line and the next

Reviewer 2: ```suggestion
| `podSelector`| A [podSelector](#podSelector) selects pods in the same namespace. |
```

| `port`| a port name or number. Must exist in a pod spec. |

> | `proxyProtocol`| Configures protocol discovery for inbound connections Supersedes the `config.linkerd.io/opaque-ports` annotation. Must be one of `unknown`,`HTTP/1`,`HTTP/2`,`gRPC`,`opaque`,`TLS`. Defaults to `unknown` if not set |

Reviewer: Period after connections, and at end of the final sentence.

Reviewer 2: ```suggestion
| `proxyProtocol`| Configures protocol discovery for inbound connections. Supersedes the `config.linkerd.io/opaque-ports` annotation. Must be one of `unknown`,`HTTP/1`,`HTTP/2`,`gRPC`,`opaque`,`TLS`. Defaults to `unknown` if not set. |
```

{{< /table >}}

### podSelector

> This is the [same labelSelector field in Kuberentes](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/label-selector/#LabelSelector).

Reviewer: speling

Reviewer 2: ```suggestion
This is the [same labelSelector field in Kubernetes](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/label-selector/#LabelSelector).
```

> All the pods that are part of this selector will be part of the server group.

Reviewer: Capitalize Server

Reviewer 2: ```suggestion
All the pods that are part of this selector will be part of the `Server` group.
```

A podSelector object must contain _exactly one_ of the following fields:

{{< table >}}
| field | value |
|-------|-------|

> | `matchExpressions` | matchExpressions is a list of label selector requirements. The requirements are ANDed. See [kubernetes reference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/label-selector/#LabelSelector) for more details |

Reviewer: - "the \[Kubernetes LabelSelector reference\]\(...\)"
- Add a period at the end of sentence.

Reviewer 2: ```suggestion
| `matchExpressions` | matchExpressions is a list of label selector requirements. The requirements are ANDed. See [the Kubernetes LabelSelector reference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/label-selector/#LabelSelector) for more details. |
```

> | `matchLabels` | matchLabels is a map of {key,value} pairs |

Reviewer: Is there a better way to describe how this means Kubernetes labels specifically?

Author: Couldn't find anything specific here. Moved the see more reference after the table as it covers information w.r.t both the fields.

{{< /table >}}

### Server Examples

A [Server](#Server) that selects over pods with a specific label, with `gRPC` as
the `proxyProtocol`.

```yaml
apiVersion: policy.linkerd.io/v1beta1
kind: Server
metadata:
  namespace: emojivoto
  name: emoji-grpc
spec:
  podSelector:
    matchLabels:
      app: emoji-svc
  port: grpc
  proxyProtocol: gRPC
```

A [Server](#Server) that selects over pods with `matchExpressions`, with `HTTP/2`
as the `proxyProtocol`, on port `8080`.

```yaml
apiVersion: policy.linkerd.io/v1beta1
kind: Server
metadata:
  namespace: emojivoto
  name: backend-services
spec:
  podSelector:
    matchExpressions:
    - {key: app, operator: In, values: [voting-svc, emoji-svc]}
    - {key: environment, operator: NotIn, values: [dev]}
  port: 8080
  proxyProtocol: "HTTP/2"
```

## ServerAuthorization

A [ServerAuthorization](#ServerAuthorization) provides a way to set client

> restrictions on a single or a group of [Server](#Server)s.

Reviewer: "on one or more"

Reviewer 2: ```suggestion
restrictions on one or more [`Server`](#Server)s.
```

### Spec

A ServerAuthorization spec must contain the following top level fields:

{{< table >}}
| field| value |
|------|-------|

> | `client`| a [client](#client) describes clients authorized to access a server. |

Reviewer: Capitalize these lines

Reviewer 2: ```suggestion
| `client`| A [client](#client) describes clients authorized to access a server. |
```

> | `server`| a [server](#server) identifies servers in the same namespace for which this authorization applies. |

Reviewer 2: ```suggestion
| `server`| A [server](#server) identifies `Servers` in the same namespace for which this authorization applies. |
```

{{< /table >}}

> ### server

Reviewer: Capitalize Server here and in the next few lines

Reviewer 2: ```suggestion
### Server
```

> A `server` object must contain _exactly one_ of the following fields:

Reviewer 2: ```suggestion
A `Server` object must contain _exactly one_ of the following fields:
```

{{< table >}}
| field | value |
|-------|-------|
| `matchExpressions` | matchExpressions is a list of label selector requirements. The requirements are ANDed. See [kubernetes reference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/label-selector/#LabelSelector) for more details |
| `matchLabels` | matchLabels is a map of {key,value} pairs |
{{< /table >}}

### client

A `client` object must contain _exactly one_ of the following fields:

{{< table >}}
| field| value |
|------|-------|
| `meshTLS`| a [meshTls](#meshTLS) is used to authorize meshed clients to access a server. Must be one of [unauthenticatedTLS](#unauthenticatedTLS),[identities](#identities),[serviceAccounts](#serviceAccounts) |
| `unauthenticated`| a boolean value that authorizes unauthenticated clients to access a server. |
{{< /table >}}

Optionally, It can also contain the `networks` field:

{{< table >}}
| field| value |
|------|-------|
| `networks`| Limits the client IP addresses to which this authorization applies. If unset, the server chooses a default (typically, all IPs or the cluster's pod network). |
{{< /table >}}

### meshTLS

> A `meshTLs` object must contain _exactly one_ of the following fields:

Reviewer 2: ```suggestion
A `meshTLS` object must contain _exactly one_ of the following fields:
```

{{< table >}}
| field| value |
|------|-------|
| `unauthenticatedTLS`| a boolean to indicate that no client identity is required for communication.This is mostly important for the identity controller, which must terminate TLS connections from clients that do not yet have a certificate. |
| `identities`| a list of proxy identity strings (as provided via MTLS) that are authorized. The `*` prefix can be used to match all identities in a domain. An identity string of `*` indicates that all authentication clients are authorized. |
| `serviceAccounts`| a [serviceAccount](#serviceAccount) authorizes clients with provided proxy identity service accounts (as provided via MTLS) |
{{< /table >}}

### serviceAccount

> A serviceAccount field must contain the following top level fields:

Reviewer 3: Since `namespace` is optional, this should read `A serviceAccount field must specify the name field`

Author: Hmm. I updated it to remove the `must` keyword and this seems to make sense after. WDYT?

{{< table >}}
| field| value |
|------|-------|
| `name`| the ServiceAccount's name. |
| `namespace`| The ServiceAccount's namespace. If unset, the authorization's namespace is used. |
{{< /table >}}

### ServerAuthorization Examples

A [ServerAuthorization](#ServerAuthorization) that allows meshed clients with

> `*.emojivoto.serviceaccount.identity.linkerd.cluster.local` proxy identity i.e all

Reviewer: "i.e."

Reviewer 2: ```suggestion
`*.emojivoto.serviceaccount.identity.linkerd.cluster.local` proxy identity i.e. all
```
